datafusion/expr/src/expr.rs

```diff
@@ -433,6 +428,22 @@ impl Cast {
     }
 }
 
+/// TryCast Expression
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct TryCast {
+    /// The expression being cast
+    pub expr: Box<Expr>,
+    /// The `DataType` the expression will yield
+    pub data_type: DataType,
+}
+
+impl TryCast {
+    /// Create a new TryCast expression
+    pub fn new(expr: Box<Expr>, data_type: DataType) -> Self {
+        Self { expr, data_type }
+    }
+}
```

Comment:
the API might be nicer if it did the boxing as if you already have a `Box` you can create the struct directly via `TryCast { expr, data_type}` 🤔 

```suggestion
impl TryCast {
    /// Create a new TryCast expression
    pub fn new(expr: Expr, data_type: DataType) -> Self {
        Self { expr: Box::new(expr), data_type }
    }
}
```